### DISCUSSION TITLE
Easier to read IEnumerable "Contains" and IDictionary "ContainsKey" and "ContainsValue" statements.

### DISCUSSION THREAD

**Discussion by mighele** [dotnet/csharplang#5708] [Language Ideas]
Hi csharplang.

I have a suggestion for C# that I think would make "Contains" statements easier to read and write.

Take the example where you are checking if a given value is in an IEnumerable. Currently you could write the following:

`if (listOfWords.Contains(targetWord)) {`

My suggestion is to be able to write this like the below:

`if (targetWord in listOfWords) {`

This could also be extended to Dictionaries. Instead of writing the following:

`if (dictionaryOfWords.ContainsKey(targetWord)) {`

You could write:

`if (targetWord in dictionaryOfWords.Keys) {`

This becomes a lot more helpful when you have multiple checks like this example from some real-world code I'm working with:

`if (!tankIds.ContainsKey(activity.Source.ID) || !tankIds.ContainsKey(activity.Destination.ID))`

Could be written as:

`if (activity.Source.ID not in tankIds.Keys || activity.Destination.ID not in tankIds.Keys) {`

I believe this makes these sort of statements easier to read as well as more intuitive to write. Often when writing this type of statement I naturally think about the problem as "is this needle in the haystack?" rather than "does the haystack contain this needle?".

**Comment by HaloFour** [comment]
`targetWord in dictionaryOfWords.Keys` would be _substantially_ less efficient than `dictionaryOfWords.ContainsKey(targetWord)`

IMO the existing methods are more than sufficient, and while they do seem to be an inverse of what we might expect that is more a matter of stylistic preference.  Putting the onus on the compiler to implement that kind of seek functionality could never be made as efficient unless the code gen was literally the same thing, but that implies that the compiler has any idea that `x.Keys` is related to `x.ContainsKey`.

**Reply by HaloFour** [reply]
If you wanted to inverse the order you could always define some extensions methods:

```cs
public static class InExtensions {
    public static bool In<T>(this T value, IList<T> list) => list.Contains(value);
    public static bool KeyIn<TKey, TValue>(this TK key, IDictionary<TKey, TValue> dictionary) => dictionary.ContainsKey(key);
}

if (word.KeyIn(dictionaryOfWords)) { ... }
```

**Reply by TahirAhmadov** [reply]
I would add that a more typical situation is the `params` scenario:
```cs
public static class InExtensions {
  public static bool In<T>(this T value, params T[] list) => list.Contains(value);
  public static bool In<T>(this T value, T t0, T t1) => object.Equals(value, t0) || object.Equals(value, t1); 
  // and a few more overloads to optimize the most common number of parameters (say, up to 5 options)
}
if(word.In("hello", "world")) { ... }
```

**Comment by huoyaoyuan** [comment]
I explained in some discussion several years earlier:

When you type a collection object first, it's fairly common to start a collection operation. But if you start with any random object, there will be less chance to start a collection operation. So it get's less support for automatic completion.

**Comment by TonyValenti** [comment]
I have a long wanted an "in" operator that would function like that and could also be equivalent to a series of "or" expressions.

Ie.
If (A in {1,2,3,5,11} P){
  ...
}
